src/metabase/search/postgres/ingestion.clj

```diff
@@ -97,17 +105,57 @@
           (t2/reducible-query (legacy-search-items-query))
           (map spec-index-reducible (keys (methods search.spec/spec)))))
 
-(defn populate-index!
-  "Go over all searchable items and populate the index with them."
-  []
-  (->> (search-items-reducible)
+(defn- batch-update! [search-items-reducible]
+  (->> search-items-reducible
        (eduction
         (comp
          (map t2.realize/realize)
          (map ->entry)
          (partition-all insert-batch-size)))
        (run! search.index/batch-update!)))
 
+(defn populate-index!
+  "Go over all searchable items and populate the index with them."
+  []
+  (batch-update! (search-items-reducible)))
+
+(defn update-index!
+  "Given a new or updated instance, create or update all the corresponding search entries."
+  [instance]
+  (when-let [updates (seq (search.spec/search-models-to-update instance))]
+    (->> (for [[search-model where-clause] updates]
+           (spec-index-reducible search-model where-clause))
+         ;; init collection is only for clj-kondo, as we know that the list is non-empty
+         (reduce u/rconcat [])
+         (batch-update!))))
+
+;; TODO think about how we're going to handle cascading deletes.
+;; Ideas:
+;; - Queue full re-index (rather expensive)
+;; - Queue "purge" (empty left join to the model) - needs special case for indexed-entity
+;; - Pre-delete hook using pre-calculated PK-based graph
+(defn delete-model!
+  "Given a deleted instance, delete all the corresponding search entries."
+  [instance]
+  (let [model (t2/model instance)
```

Comment:
should we assert instance is a toucan instance?

Reply:
Interesting idea. Since it'll just be a no-op if not, I don't think it's necessary. Since we only plan to wire this up to toucan hooks as well it seems fine as is.

Reply:
I find that kind of asserts are quite useful for development purposes e.g: to make sure we use these functions at the right place. 

Speaking of which maybe adding a type `(ms/InstaceOf ToucanModel)` is enough.

up to you tho